### DISCUSSION TITLE
[Contract Verification] Verify plugin reads incorrect number of constructor arguments

### DISCUSSION THREAD

**Discussion by boyuanx** [zkSync-Community-Hub/zksync-developers#389] [Contract Verification] [answered]
### Environment

Mainnet

### Issue Type

- [X] Plugin
- [ ] Block Explorer

### Contract Address

0xDe7a77Dd6B283429522bBdfD268E10A3F4E6950A

### Compiler Type

Multi-part contract

### zkSolc Version

1.4.0

### Solc Version

0.8.24

### Contract Name

TTUV2BeaconManager

### Contract Code

```solidity
// SPDX-License-Identifier: AGPL v3

pragma solidity ^0.8.20;

import {UpgradeableBeacon} from "@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
import {IVersionable} from "../interfaces/IVersionable.sol";

/**
 * @title TTUV2BeaconManager
 * @author Jack Xu @ EthSign
 * @dev This contract manages the upgradeable beacons that we use to seamlessly
 * upgrade TokenTableUnlocker, TTFutureToken, and TTTrackerToken on behalf of
 * our users in the future.
 *
 * This contract should be deployed using TTUDeployer.
 */
contract TTUV2BeaconManager is Ownable, IVersionable {
    UpgradeableBeacon public immutable unlockerBeacon;
    UpgradeableBeacon public immutable futureTokenBeacon;
    UpgradeableBeacon public immutable trackerTokenBeacon;

    constructor(
        address unlockerImpl,
        address futureTokenImpl,
        address trackerTokenImpl
    ) Ownable(_msgSender()) {
        unlockerBeacon = new UpgradeableBeacon(unlockerImpl, address(this));
        futureTokenBeacon = new UpgradeableBeacon(
            futureTokenImpl,
            address(this)
        );
        trackerTokenBeacon = new UpgradeableBeacon(
            trackerTokenImpl,
            address(this)
        );
    }

    function upgradeUnlocker(address newImpl) external onlyOwner {
        unlockerBeacon.upgradeTo(newImpl);
    }

    function upgradeFutureToken(address newImpl) external onlyOwner {
        futureTokenBeacon.upgradeTo(newImpl);
    }

    function upgradePreviewToken(address newImpl) external onlyOwner {
        trackerTokenBeacon.upgradeTo(newImpl);
    }

    function version() external pure returns (string memory) {
        return "2.0.1";
    }
}

```

### Constructor Arguments

`[0xa6b355a2Ef8733FbD907C364115c01046e3EC80a, 0xC66F095E645E12892b8637830810a800dCF691A7, 0xe225A7B9FeE734e5B2Ce18DF926d2892f1adCD08]`

### Repo Link (Optional)

https://github.com/EthSign/tokentable-v2-evm

### Additional Details

Error message:
`Error in plugin @matterlabs/hardhat-zksync-verify: The number of constructor arguments you provided (3) does not match the number of constructor arguments the contract has been deployed with (2).`

**Comment by mpopovac-txfusion** [comment, marked as answer]
Hey @boyuanx , we have tried using our `hardhat-zksync` plugins and everything worked. It seems that the community plugin you are using for deploying (`hardhat-deploy`) doesn't handle the tx receipt in a proper way on zkSync, meaning it doesn't return the right address for the given contract and instead returns the address from one of the deployed upgradable beacons. We will try to contribute to that community plugin, but until then, you can use ours.
Also, you can check the logs of the deployed transaction; there, in one of them, you can find the correct address that you can use for your verification.

**Reply by boyuanx** [reply]
Thank you. I have both `hardhat-zksync-deploy` and `hardhat-deploy` installed and if I'm understanding you correctly, are you saying I've been using the community plugin instead of the zkSync plugin? Is there a way to use one over the other? I did mark zkSync networks with `zksync: true` in Hardhat config.

**Reply by mpopovac-txfusion** [reply]
When you are using `hre.deployments`, in the background you are calling extensions from the `hardhat-deploy` plugin. For `hardaht-zksync-deploy` you can use `hre.deployer`.

Here is the comparison between using `hardhat-deploy` and `hardhat-zksync-deploy` plugin for deploying `TTUV2BeaconManager`:

Current code (`hardhat-deploy`):
```
const beaconManagerDeploymentResult = await deploy('TTUV2BeaconManager', {
        from: deployer,
        log: true,
        args: [
            unlockerDeploymentResult.address,
            ftDeploymentResult.address,
            ttDeploymentResult.address
        ],
        waitConfirmations: 1
    })
```
    
`hardhat-zksync-deploy`:

```
const beaconManagerArtifact = await hre.deployer.loadArtifact('TTUV2BeaconManager')
const beaconManagerDeploymentResult = await hre.deployer.deploy(beaconManagerArtifact, [
      unlockerDeploymentResult.address,
      ftDeploymentResult.address,
      ttDeploymentResult.address
])
```

**Comment by dgeyston** [comment]
The error indicates a mismatch between the number of constructor arguments provided during deployment and the number expected by the contract. Ensure consistency between deployment parameters and contract expectations.

**Comment by mariyokaveen** [comment]
The error occurs because the contract constructor expects 3 arguments, but it was deployed with only 2 arguments.

**Comment by bryanbrega** [comment]
The provided contract code is for TTUV2BeaconManager, which manages the upgradeable beacons used to upgrade TokenTableUnlocker, TTFutureToken, and TTTrackerToken on behalf of users. The contract is deployed using TTUDeployer and includes functions to upgrade the various beacons. The contract is written in Solidity version 0.8.20 and uses the OpenZeppelin UpgradeableBeacon and Ownable contracts. The contract implements the IVersionable interface and returns version "2.0.1". The contract is deployed on Mainnet at address 0xDe7a77Dd6B283429522bBdfD268E10A3F4E6950A and was compiled using zkSolc version 1.4.0.
